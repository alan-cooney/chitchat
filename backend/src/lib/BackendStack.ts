import { Stack, Construct, StackProps } from "@aws-cdk/core";
import {
  AttributeType,
  BillingMode,
  StreamViewType,
  Table,
} from "@aws-cdk/aws-dynamodb";
import { AuthorizationType, GraphqlApi, Schema } from "@aws-cdk/aws-appsync";
import { join } from "path";
import { spawnSync } from "child_process";

export default class BackendStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Code-generate the Appsync-suitable schema before continuing
    spawnSync("graphql-codegen");

    const table = new Table(this, "table", {
      billingMode: BillingMode.PAY_PER_REQUEST,
      partitionKey: {
        name: "pk",
        type: AttributeType.STRING,
      },
      pointInTimeRecovery: true,
      sortKey: {
        name: "sk",
        type: AttributeType.STRING,
      },
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
    });

    const api = new GraphqlApi(this, "api", {
      name: id,
      schema: Schema.fromAsset(
        join(__dirname, "../../dist/graphql-codegen/combined.graphql")
      ),
      authorizationConfig: {
        defaultAuthorization: {
          authorizationType: AuthorizationType.IAM,
          // openIdConnectConfig: {
          //   oidcProvider: id.endsWith("Prox")
          //     ? "https://chitchat-prod.eu.auth0.com"
          //     : "https://chitchat-dev.eu.auth0.com",
          // },
        },
      },
    });

    api.addDynamoDbDataSource("tableDataSource", table);
  }
}
