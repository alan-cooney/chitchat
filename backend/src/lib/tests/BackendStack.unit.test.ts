import "@aws-cdk/assert/jest";
import { App } from "@aws-cdk/core";
import BackendStack from "../BackendStack";

const app = new App();
const stack = new BackendStack(app, "test");

jest.mock("child_process");

it("creates a dynamodb table", () => {
  expect(stack).toHaveResource("AWS::DynamoDB::Table", {
    BillingMode: "PAY_PER_REQUEST",
    KeySchema: [
      {
        AttributeName: "pk",
        KeyType: "HASH",
      },
      {
        AttributeName: "sk",
        KeyType: "RANGE",
      },
    ],
    PointInTimeRecoverySpecification: {
      PointInTimeRecoveryEnabled: true,
    },
    StreamSpecification: {
      StreamViewType: "NEW_AND_OLD_IMAGES",
    },
  });
});

it("creates the API schema", () => {
  expect(stack).toHaveResource("AWS::AppSync::GraphQLSchema");
});
