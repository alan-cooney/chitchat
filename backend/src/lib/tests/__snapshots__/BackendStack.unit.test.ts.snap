// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates resources matching the snapshot 1`] = `
Object {
  "Resources": Object {
    "apiC8550315": Object {
      "Properties": Object {
        "AuthenticationType": "API_KEY",
        "Name": "test",
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "apiDefaultApiKey6AB8D7C4": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "apiSchema0EA92056": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "Definition": "schema {
  query: Query
  mutation: Mutation
  subscription: Subscription
}

input AddMessageInput {
  id: ID!
  conversationID: ID!
  text: String!
}

input AddMessageSubscriptionInput {
  conversationID: ID!
}

input AddParticipantInput {
  conversationID: ID!
  userID: ID!
}

input AddedParticipantSubscriptionInput {
  userID: ID!
}

type Conversation implements Node {
  id: ID!
  participants: [User!]!
  messages(limit: Int!, nextKey: String!): [Message!]!
}

input CreateConversationInput {
  id: ID!
}

type Message implements Node {
  id: ID!
  conversation: Conversation!
  text: String!
}

type Mutation {
  createConversation(input: CreateConversationInput!): Conversation!
  addParticipant(input: AddParticipantInput!): Conversation!
  addMessage(input: AddMessageInput!): Message!
}

interface Node {
  id: ID!
}

type Participant {
  conversation: Conversation!
  user: User!
}

type Query {
  user: User!
}

type Subscription {
  addedParticipant(input: AddedParticipantSubscriptionInput!): Conversation! @aws_subscribe(mutations: [\\"addParticipant\\"])
  addedMessage(input: AddMessageSubscriptionInput!): Message! @aws_subscribe(mutations: [\\"addMessage\\"])
}

type User implements Node {
  id: ID!
  givenName: String!
  familyName: String!
  email: String!
  phone: String!
  conversations(limit: Int!, nextKey: String!): [Conversation!]!
}
",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
    "table8235A42E": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
